homeassistant/config_entries.py

```diff
@@ -63,20 +64,37 @@
 
 SAVE_DELAY = 1
 
-# The config entry has been set up successfully
-ENTRY_STATE_LOADED = "loaded"
-# There was an error while trying to set up this config entry
-ENTRY_STATE_SETUP_ERROR = "setup_error"
-# There was an error while trying to migrate the config entry to a new version
-ENTRY_STATE_MIGRATION_ERROR = "migration_error"
-# The config entry was not ready to be set up yet, but might be later
-ENTRY_STATE_SETUP_RETRY = "setup_retry"
-# The config entry has not been loaded
-ENTRY_STATE_NOT_LOADED = "not_loaded"
-# An error occurred when trying to unload the entry
-ENTRY_STATE_FAILED_UNLOAD = "failed_unload"
-
-UNRECOVERABLE_STATES = (ENTRY_STATE_MIGRATION_ERROR, ENTRY_STATE_FAILED_UNLOAD)
+
+class EntryState(Enum):
```

Comment:
This will break any custom integration with a configuration flow out there.

Reply:
:/ No experience with those. Ideas what to do about that / how to proceed?

Reply:
That, in general, depends a bit on the change. While we don't support custom integrations, that community uses them a lot, and making a breaking change like that is something to consider/mitigate.

So, normally we try to drop a warning in the logs, but that would be hard with constants, so we often keep them around in a backward-compatible matter for some time.

We could write a dev blog about it, but honestly, I'm not sure how much of an impact that will have.

Reply:
I clicked through all integrations currently visible in HACS, searched for "state" in them, and skimmed the results: noticed two integrations that would be broken by this:
- HACS itself: https://github.com/hacs/integration/blob/fb9b64f4eae869dd97215beb5be21b14529b4801/custom_components/hacs/operational/remove.py#L14 -- breakage doesn't seem bad, sensor unload wouldn't happen on disable
- hass-tdameritrade: https://github.com/prairiesnpr/hass-tdameritrade/blob/fc4013a56aec8ae8dac287a1097f8910f691adb6/custom_components/tdameritrade/__init__.py#L133 -- breakage worse, binary_sensor and sensor platforms wouldn't load

If it's limited to those and HACS itself isn't even that badly affected, to me this would sound acceptable, naturally with a breaking change notice attached.

Reply:
Maybe use a data-structure that will still string-ify to the original value?

Reply:
I don't think that would help with comparing it against a string, such as is the case in the two linked ones above. But this added into the enum would work around it and thus in those two:
```python
    def __eq__(self, other):
        return self is other or self.value == other
```
That's pretty much directly against PEP 435 though, see end of [Comparisons](https://www.python.org/dev/peps/pep-0435/#comparisons) and [IntEnum](https://www.python.org/dev/peps/pep-0435/#intenum) chapters there.

Reply:
Or rather something like this:
```python
    def __eq__(self, other):
        if self is other:
            return True
        if isinstance(other, str):
            # Log deprecation warning here
            return self.value == other
        return self == other
```

Reply:
Implemented the above.

Reply:
Indeed, a search through all repos in HACS, seems to have limited result (as in, barely any) as you indeed concluded already as well. HACS we can adjust quite easy.

I would say, lets not pollute our repo for the low of an impact.

@ludeeus Considering you are the HACS expert around here, what do you think?

Reply:
For HACS itself, I can adjust that long before 2021.6

Reply:
Given the low impact on custom integrations and the cleaner code that we get out of it, I say let's move forward.